app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt

```diff
@@ -759,6 +778,59 @@ class BrowserTabFragment :
         }
     }
 
+    private fun extractUrlFromTrackingLink(initialUrl: String) {
+        context?.let {
+
+            val client = urlExtractingWebViewClient.get()
+
+            val listener = object : UrlExtractionListener {
+                override fun onUrlExtractionError() {
+                    Timber.d("Tracking link detection: Extraction error! Loading initial URL: $initialUrl")
+                    webView?.loadUrl(initialUrl)
+                    destroyUrlExtractingWebView()
+                }
+
+                override fun onUrlExtracted(extractedUrl: String?) {
+                    val destinationUrl: String
+
+                    if (extractedUrl != null) {
+                        destinationUrl = extractedUrl
+                        viewModel.updateLastTrackingLink(initialUrl)
+                        Timber.d("Tracking link detection: Success! Loading extracted URL: $destinationUrl")
+                    } else {
+                        destinationUrl = initialUrl
+                        Timber.d("Tracking link detection: Failed! Loading initial URL: $destinationUrl")
+                    }
+
+                    requireActivity().runOnUiThread {
+                        webView?.loadUrl(destinationUrl)
+                        destroyUrlExtractingWebView()
+                    }
+                }
+            }
+
+            client.urlExtractionListener = listener
+
+            Timber.d("Tracking link detection: Creating WebView for URL extraction")
+            urlExtractingWebView = UrlExtractingWebView(requireContext(), client, urlExtractorUserAgent.get(), urlExtractor.get())
+
+            urlExtractingWebView?.urlExtractionListener = listener
+
+            Timber.d("Tracking link detection: Loading tracking URL for extraction")
+            urlExtractingWebView?.loadUrl(initialUrl)
+        }
+    }
+
+    private fun destroyUrlExtractingWebView() {
```

Comment:
I would extract this into an extension function inside `UrlExtractingWebView` class.

`fun UrlExtractingWebView.destroy() { .... } `